File: client/common/serializer.go
package common

import (
	"encoding/binary"
	"errors"
)

const SendBetBatchOpcode = 1
const BetPayloadSize = 144

type Serializer struct{}

func (s *Serializer) int16ToBytes(n uint16) []byte {
	b := make([]byte, 2)
	binary.BigEndian.PutUint16(b, uint16(n))
	return b
}

func (s *Serializer) int32ToBytes(n uint32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, uint32(n))
	return b
}

func (s *Serializer) deserializeOpcode(data []byte) uint16 {
	return binary.BigEndian.Uint16(data)
}

func (s *Serializer) SerializeBets(bets []Bet, clientID uint16) ([]byte, error) {

	// Convert "header" values to bytes and append at first 6 message bytes
	// (2 bytes for opcode, 2 bytes for clientID, 2 bytes for number of bets)
	opcodeBytes := s.int16ToBytes(SendBetBatchOpcode)
	data := append(opcodeBytes, s.int16ToBytes(clientID)...)
	data = append(data, s.int16ToBytes(uint16(len(bets)))...)

	for _, bet := range bets {

		betNumberBytes := s.int16ToBytes(bet.Number)
		documentBytes := s.int32ToBytes(bet.Document)

		// Convert dynamic-length strings to bytes and calculate its length
		firstNameBytes := []byte(bet.FirstName)
		firstNameLenBytes := s.int16ToBytes(uint16(len(firstNameBytes)))

		lastNameBytes := []byte(bet.LastName)
		lastNameLenBytes := s.int16ToBytes(uint16(len(lastNameBytes)))

		// Assume BirthDate is always a 10 byte string
		birthDateBytes := []byte(bet.BirthDate)

		// Combine all byte slices into one and add padding if necesary
		bet_bytes := append(documentBytes, betNumberBytes...)
		bet_bytes = append(bet_bytes, birthDateBytes...)
		bet_bytes = append(bet_bytes, firstNameLenBytes...)
		bet_bytes = append(bet_bytes, firstNameBytes...)
		bet_bytes = append(bet_bytes, lastNameLenBytes...)
		bet_bytes = append(bet_bytes, lastNameBytes...)

		if len(bet_bytes) < BetPayloadSize {
			padding := make([]byte, BetPayloadSize-len(bet_bytes))
			bet_bytes = append(bet_bytes, padding...)
		} else if len(bet_bytes) > BetPayloadSize {
			return nil, errors.New("bet fields size is too large")
		}
		data = append(data, bet_bytes...)
	}

	return data, nil
}

File: server/common/serializer.py
from .utils import Bet


class Serializer:
    
    def serialize_opcode(self, opcode:int) -> bytes:
        return self.serialize_int_to_bytes(opcode, 2)
    
    def deserialize_int_to_bytes(self, bytes:bytes) -> int:
        return int.from_bytes(bytes, byteorder='big')
    
    def deserialize_bet(self, betBytes:bytes, agency_number:int) -> Bet:
        document = int.from_bytes(betBytes[0:4], byteorder='big')
        number = int.from_bytes(betBytes[4:6], byteorder='big')
        birthdate = betBytes[6:16].decode('utf-8')
        first_name_len =  int.from_bytes(betBytes[16:18], byteorder='big')
        first_name = betBytes[18:18+first_name_len].decode('utf-8')
        last_name_len = int.from_bytes(betBytes[18+first_name_len: 20+first_name_len], byteorder='big')
        last_name = betBytes[20+first_name_len:20+first_name_len+last_name_len].decode('utf-8')
        
        return Bet(
            str(agency_number),
            first_name,
            last_name,
            str(document),
            birthdate,
            str(number)
        )
            
    def serialize_int_to_bytes(self, number:int, size:int) -> bytes:
        return number.to_bytes(size, byteorder='big')